### PR TITLE
Add support for ENCODE BigBed types

### PR DESCRIPTION
**Pull request by zoldello**
The purpose of this PR is to add support to for ENCODE BigBed types. These types have columns that differs the standard BigBed types.

### CODE REVIEW

**Review by jackh726 — CHANGES_REQUESTED** [review]
I left a few nits. But my biggest concern with this PR is that it ends up being a lot of boilerplate, so I propose an alternative API here.

First, you have parsing functions of the signature:
```
interface ParseFunction {
  (chrom: string, start: number, end: number, rest: string) => T;
}
```
First, the signature of `readBigBedData` (and `streamBigBedData`) changes to include an extra parameter `T` and argument `restParser: ParseFunction<T> = parseBigBedDefault`.

Then, the parsing for the different bed formats end up in different functions (outside the class). Example of the default:
```
function parseBigBedDefault(chrom: string, start: number, end: number, rest: string): BigBedData {
  const entry: BigBedData = {
    chr: chrom,
    start: startBase,
    end: endBase
  }
  let tokens = rest.split("\t");
  if (tokens.length > 0) {
    entry.name = tokens[0];
  }
  ...
  return entry;
}
```

Then, to read the different types of bed formats you can just do something like:
```
const regularData = await reader1.readBigBedData("chr21", 33_037_487, "chr21", 33_047_461);
const narrowPeakData = await reader1.readBigBedData("chr21", 33_037_487, "chr21", 33_047_461, parseBigBedNarrowPeak);
```

CI should be fixed on next push too.

**Comment by zoldello** [comment]
@jackh726  
I am Phil. I am a Developer at Stanford, in Cherry’s Lab. I work on the ENCODE project. Thanks for the feedback and I am working to address them. You made great points. 

I notice the boilerplate too but was hesitant to touch signatures of existing methods. Your feedback on removing boilerplate code will result in breaking changes technically. For example, the signatures of readBigBedData will get an extra parameter. That means the "major" part of the semantic version (major.minor.patch); will have to go up (which has ramifications). Is your team okay with that? Is your team also okay with adding eslint? It's great at catching format errors. Do you have a contributor.md file or coding standards guidelines?

**Comment by jackh726** [comment]
> @jackh726
> I am Phil. I am a Developer at Stanford, in Cherry’s Lab. I work on the ENCODE project. Thanks for the feedback and I am working to address them. You made great points.
> 
> I notice the boilerplate too but was hesitant to touch signatures of existing methods. Your feedback on removing boilerplate code will result in breaking changes technically. For example, the signatures of readBigBedData will get an extra parameter. That means the "major" part of the semantic version (major.minor.patch); will have to go up (which has ramifications). Is your team okay with that? Is your team also okay with adding eslint? It's great at catching format errors. Do you have a contributor.md file or coding standards guidelines?

It would be a new parameter; but it would have a default, so it would *not* be a breaking change. For style errors, we generally use prettier for other projects, but don't have it enabled for this project. I really only pointed out the style nits since they didn't quite match the rest of the code, but they aren't a *huge* deal. No contribuer.md, but that would be useful!

**Comment by zoldello** [comment]
@jackh726 
I got busy with preparing for a release and some other things. I and other lab members are not familiar with Typescript and its generic-implementation. I have tried two approaches to solve this- One my idea and the second trying yours. But have weird errors.

**Approach #1:  Pass decode function**
https://gist.github.com/zoldello/ac29a20bcf60134c321830ed6d80b120

Beginning in  line 298, I am trying to pass "decodeBedData" function (will use similar ones for other big bed types maybe later), I get that the value of T is unknown. I tried making "decodeBedData" generic but that lead to BigBedData interface being recognized as generic. Do you have any educated guess on the issue? In general, the error is:

`(parameter) decodeFunction: DecodeFunction<T>
Type '(data: ArrayBuffer, filterStartChromIndex: number, filterStartBase: number, filterEndChromIndex: number, filterEndBase: number, chromDict: string[]) => BigBedData[]' is not assignable to type 'DecodeFunction<T>'.
  Type 'BigBedData[]' is not assignable to type 'T[]'.
    Type 'BigBedData' is not assignable to type 'T'.
      'T' could be instantiated with an arbitrary type which could be unrelated to 'BigBedData'.ts(2322)
`


**Approach #2: Your approach with modification**
https://gist.github.com/zoldello/715e3c1b4b20c2c74a3122d62102578d

I tried your approach but using "type" instead of "interface". The main code begins at line 271. Do you have any educated guesses what could be wrong? I have an similar error as Approach 1

**Comment by jackh726** [comment]
@zoldello No worries. Been playing around with this a bit. This is a little finicky because we want typescript to realize that if we don't pass a rest parser function, we will be returning `BigBedData`, but if we do, we will be returning `T`. I had assumed that having the default parameter there would make Typescript assume that `T` defaults to `BigBedData` if `restParser` isn't passed (I even tried `T = BigBedData` but that didn't work).

I did find a signature that works though, which uses function overloading (for approach 2). It did require one `as any` unfortunately:
```
    async readBigBedData(startChrom: string, startBase: number, endChrom: string, endBase: number): Promise<Array<BigBedData>>;
    async readBigBedData<T>(startChrom: string, startBase: number, endChrom: string, endBase: number, restParser: ParseFunction<T>): Promise<Array<T>>;
    async readBigBedData<T>(startChrom: string, startBase: number, endChrom: string, endBase: number, restParser?: ParseFunction<T>): Promise<Array<(T | BigBedData)>> {
        return this.readData(startChrom, startBase, endChrom, endBase, 
            (await this.getHeader()).common!.fullIndexOffset, decodeBedData, restParser || parseBigBed as any);
    }
```

A couple points to mention: 
* I don't think we need to provide special parsing for bigWigs or for zoom data. These formats are fixed. BigBeds are a little bit different, because we want to provide a nice wrapper for parsing the "rest" data of each entry.
* A similar overloading approach must be taken as above for `streamBigBedData` to ensure that we don't have a breaking change to the API there either.
* Having a bunch of functions like `readBigBedDataNarrowPeak` that just call to `readData` with the different parse functions is not necessarily bad, but I would rather leave that out of the API surface, least for now. It's easy enough to export the parse functions and have users pass that in for any format that's not the default right now.
* Can you move the parse functions to a separate file and make sure they are all exported? Also, probably the interfaces for all associated data too. The `BigWigData` and `BigBedData` interfaces can stay.

Here is a gist with some tweaks: https://gist.github.com/jackh726/815fa96ca062cb3e4f1305c924375986
And tests that show how it would be used:
```
    it("should read unzoomed bigbed data", async () => {
        const loader = new AxiosDataLoader(`http://localhost:8001/${testBBFilename}`, Axios.create());
        const reader = new BigWigReader(loader);
        const data = await reader.readBigBedData("chr21", 10_000_000, "chr21", 20_000_000);
        expect(data.length).toBe(46);
        expect(data[0].chr).toBe("chr21");
        expect(data[0].start).toBe(9_928_613);
        expect(data[0].end).toBe(10_012_791);
        expect(data[0].exons!.length).toBe(22);
        expect(data[0].exons![0].start).toBe(9_928_613);
        expect(data[0].exons![0].end).toBe(9_928_911);
    });

    it("should read unzoomed bigbed data passing parser", async () => {
        const loader = new AxiosDataLoader(`http://localhost:8001/${testBBFilename}`, Axios.create());
        const reader = new BigWigReader(loader);
        const data = await reader.readBigBedData("chr21", 10_000_000, "chr21", 20_000_000, parseBigBed);
        expect(data.length).toBe(46);
        expect(data[0].chr).toBe("chr21");
        expect(data[0].start).toBe(9_928_613);
        expect(data[0].end).toBe(10_012_791);
        expect(data[0].exons!.length).toBe(22);
        expect(data[0].exons![0].start).toBe(9_928_613);
        expect(data[0].exons![0].end).toBe(9_928_911);
    });

    it("should read unzoomed narrow peak bigbed data", async () => {
        const loader = new AxiosDataLoader(`http://localhost:8001/${testBBNarrowPeakFilename}`, Axios.create());
        const reader = new BigWigReader(loader);
        const data  = await reader.readBigBedData("chr21", 33_037_487, "chr21", 33_047_461, parseBigBedNarrowPeak)
        expect(data.length).toBe(2);
        expect(data[0].chr).toBe("chr21");
        expect(data[0].start).toBe(33_039_438);
        expect(data[0].end).toBe(33_039_560);
        expect(data[0].name).toBe('chr21.1228');
        expect(data[0].score).toBe(577);
        expect(data[0].strand).toBe('.');
        expect(data[0].signalValue).toBe(0);
        expect(data[0].pValue).toBe(1);
        expect(data[0].qValue).toBe(-1);
        expect(data[0].peak).toBe(84);        
    });
```

File: src/bigwig/BigWigReader.ts [hunk 1]
```diff
@@ -353,26 +359,88 @@ async function loadLeafNodesForRPNode(bufferedLoader: BufferedDataLoader, little
...
  * @param filterEndChromIndex ending chromosome index used for filtering
  * @param filterEndBase ending base used for filtering
  * @param chromDict dictionary of indices used by the file to chromosome names, conveniently stored as an array.
+ * @param [restParser] Parse for getting data
  */
-function decodeBedData(data: ArrayBuffer, filterStartChromIndex: number, filterStartBase: number, filterEndChromIndex: number,
-        filterEndBase: number, chromDict: Array<string>): Array<BigBedData> {
-    const decodedData: Array<BigBedData> = [];
+function decodeBedData<T>(data: ArrayBuffer, filterStartChromIndex: number, filterStartBase: number, filterEndChromIndex: number,
+    filterEndBase: number, chromDict: Array<string>, restParser?: ParseFunction<T>): Array<T> {
```
Comment (jackh726) [inline src/bigwig/BigWigReader.ts:439]: I would expect that `restParser` would not be optional.

File: src/bigwig/BigWigReader.ts [hunk 2]
```diff
@@ -51,6 +51,8 @@ interface RPLeafNode {
     dataSize: number;
 }
 
+type ParseFunction<T> = (chrom: string, startBase: number, endBase: number, rest: string, binaryParser?: BinaryParser) => T;
```
Comment (jackh726) [inline src/bigwig/BigWigReader.ts:54]: Don't need the `binaryParser` argument here

Comment (jackh726) [inline src/bigwig/BigWigReader.ts:54]: Can you export this?

File: src/bigwig/BigWigReader.ts [hunk 3]
```diff
@@ -237,9 +243,9 @@ export class BigWigReader {
      * @param treeOffset Location of the R+ tree that stores the data we're interested.
      * @param decodeFunction 
      */
-    private async loadData<T>(startChrom: string, startBase: number, endChrom: string, endBase: number, 
-                treeOffset: number, streamMode: boolean, decodeFunction: DecodeFunction<T>, 
-                loadFunction: LoadFunction<T>): Promise<void> {
+    private async loadData<T>(startChrom: string, startBase: number, endChrom: string, endBase: number,
+        treeOffset: number, streamMode: boolean, decodeFunction: DecodeFunction<T>,
+        loadFunction: LoadFunction<T>, restParser?: ParseFunction<T>): Promise<void> {
```
Comment (jackh726) [inline src/bigwig/BigWigReader.ts:248]: Oh, I forgot that this gets called for both bigBeds and bigWigs. An alternative to make this *not* change the function signature is to make `decodeBedData` actually be a function with the `restParser` argument that returns the actual decode function.

Then, in `readBigBedData` you pass `decodeBedData(restParser || parseBigBed as any)` for the decode function.

File: src/bigwig/index.ts [hunk 4]
```diff
@@ -1,2 +1,3 @@
 export { HeaderData, FileType, CommonHeader, ZoomLevelHeader, BWTotalSummary, ChromTree } from "./BigWigHeaderReader";
+export { BigBedDataNarrowPeak, BigBedDataBroadPeak, BigBedDataMethyl, BigBedDataTssPeak } from "./encodeBigBed";
 export { BigWigData, BigBedData, BigZoomData, BigWigReader } from "./BigWigReader";
```
Comment (jackh726) [inline src/bigwig/index.ts:3]: Can you also make sure the parse functions are exported?

Comment (jackh726) [reply src/bigwig/index.ts:3]: And the `ParseFunction` type.

**Review by jackh726 — CHANGES_REQUESTED** [review]
This looks great! A couple small things then I think this is good to merge! Thanks for working on this 🙂

**Comment by zoldello** [comment]
@jackh726 I made an update to master without knowing about your feedback. I will make sure my next one has your changes. I made restParser nullable and binaryParser optional because of BigWig. I will see if I can not do these based on your feedback

File: src/bigwig/BigWigReader.ts (hunk 3 above)
Comment (zoldello) [reply src/bigwig/BigWigReader.ts:248]: @jackh726 This is tricky. ` decodeBedData(restParser || parseBigBed as any)` can be like a closure- a function that returns a function. But I am not sure how to get the other variables that the decode function need (like filterStartChromIndex) in this approach as only the `restParser || parseBigBed as any` is available initial setup in `readBigBedData`. It also gets tricky in that `decodeZoomData` does not used restParser.

File: src/bigwig/BigWigReader.ts (hunk 1 above)
Comment (zoldello) [reply src/bigwig/BigWigReader.ts]: Done

File: src/bigwig/BigWigReader.ts (hunk 2 above)
Comment (zoldello) [reply src/bigwig/BigWigReader.ts]: Done

Comment (zoldello) [reply src/bigwig/BigWigReader.ts]: Done.

File: src/bigwig/BigWigReader.ts (hunk 3 above)
Comment (zoldello) [reply src/bigwig/BigWigReader.ts]: I used an interface as a function seemed verbose. Let me know if you prefer a function.

File: src/bigwig/index.ts (hunk 4 above)
Comment (zoldello) [reply src/bigwig/index.ts]: Done.

File: src/bigwig/BigWigReader.ts [hunk 5]
```diff
@@ -142,11 +149,13 @@ export class BigWigReader {
...
      */
-    async readBigBedData(startChrom: string, startBase: number, endChrom: string, 
-            endBase: number): Promise<Array<BigBedData>> {
-        return this.readData<BigBedData>(startChrom, startBase, endChrom, endBase, 
-            (await this.getHeader()).common!.fullIndexOffset, decodeBedData);
+    async readBigBedData(startChrom: string, startBase: number, endChrom: string, endBase: number): Promise<Array<BigBedData>>;
+    async readBigBedData<T>(startChrom: string, startBase: number, endChrom: string, endBase: number, restParser: ParseFunction<T>): Promise<Array<T>>;
+    async readBigBedData<T>(startChrom: string, startBase: number, endChrom: string, endBase: number, restParser?: ParseFunction<T>): Promise<Array<(T | BigBedData)>> {
+        return this.readData(startChrom, startBase, endChrom, endBase,
+            (await this.getHeader()).common!.fullIndexOffset, { decodeFunction: decodeBedData, restParser: restParser || parseBigBed as any });
```
Comment (jackh726) [inline src/bigwig/BigWigReader.ts:158]: This really doesn't end up being any different than a separate optional argument...you still end up having to pipe it through.

**Review by jackh726 — CHANGES_REQUESTED** [review]
This isn't quite what I was expecting. I was expecting the `decodeBedData` function to look something like:
```
const decodeBedData = <T>(restParser: ParseFunction<T>) => (data: ArrayBuffer, filterStartChromIndex: number, filterStartBase: number, filterEndChromIndex: number,
    filterEndBase: number, chromDict: Array<string>, ): Array<T> = {}
```

File: src/bigwig/BigWigReader.ts (hunk 5 above)
Comment (zoldello) [reply src/bigwig/BigWigReader.ts]: Wow. I have never seen anyone use an arrow function to make a closure. Thanks for the insights and patience. They were very helpful. Hopefully now things are perfect.

Comment (jackh726) [reply src/bigwig/BigWigReader.ts]: I personally prefer this syntax, but you could also do the same thing with:
```
function decodeBedData<T>(restParser: ParseFunction<T>) {
  return function(...): T[] {
  }
}
```

File: src/bigwig/encodeBigBed.ts [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+    qValue?: number,
+}
+
+export interface BigBedDataMethyl {
+    chr: string,
+    start: number,
+    end: number,
+    name?: string,
+    score?: number,
+    strand?: string,     // + or - or . for unknown
```
Comment (jackh726) [inline src/bigwig/encodeBigBed.ts:41]: nit: this comment is still at the end of the line

File: test/BigWigReader.test.ts [hunk 7]
```diff
@@ -1,12 +1,21 @@
 import Axios from "axios";
+import { exception } from "console";
```
Comment (jackh726) [inline test/BigWigReader.test.ts:2]: extra import here

**Review by jackh726 — APPROVED** [review]
A couple small points, but I'm not going to block merging for them.

Thanks for putting up with the review here! I personally feel like this final result ends up being much cleaner and is much more extensible too.

**Comment by cmdcolin** [comment]
Hi there, I am just a humble follower of this project (I actually really like this codebase, it ends up being simpler than some of our parsers), but one thing that is maybe interesting is that bigbed files contain "autoSql" in their header that makes them somewhat self describing


This means that you potentially do not need to add special parsers for these datatypes, as they will be encoded in the file.

I wrote https://github.com/GMOD/bbi-js/ and https://github.com/GMOD/bed-js to try to solve these problems. Note that even these parsers do not go so far as to parse the features out into exons etc as that is sort of application dependent how you represent your gene features but it is just an idea

**Comment by cmdcolin** [comment]
You can use "bigBedInfo" command line tool to inspect the autoSql of a given file. 

Example

```
testbb-tss.bigbed
version: 4
isCompressed: yes
isSwapped: 0
itemCount: 23,645
primaryDataSize: 1,179,714
zoomLevels: 0
chromCount: 24
as:
table tss_peak
"Tss peak coverage. Browser extensible data BED6+5."
    (
    string  chrom;       "Reference sequence chromosome (or contig, scaffold, etc.)"
    uint    chromStart;  "Start position in chromosome"
    uint    chromEnd;    "End position in chromosome"
    string  name;        "Name of item"
    uint    score;       "Score from 0-1000"
    char[1] strand;      "+ or - for strand"
    float   count;       "Count of reads mapping to this peak"
    string  gene_id;     "Gene identifier"
    string  gene_name;   "Gene name"
    string  tss_id;      "TSS identifier"
    lstring peak_cov;    "base by base read coverage of the first 512 bases of peak"
    )
basesCovered: 2,202,031
meanDepth (of bases covered): 1.001253
minDepth: 1.000000
maxDepth: 2.000000
std of depth: 0.035375

```

Also note, the @gmod/bed is in some ways sort of of excessive, can likely get by with a simpler approach

**Comment by jackh726** [comment]
> Hi there, I am just a humble follower of this project (I actually really like this codebase, it ends up being simpler than some of our parsers), but one thing that is maybe interesting is that bigbed files contain "autoSql" in their header that makes them somewhat self describing
> 
> This means that you potentially do not need to add special parsers for these datatypes, as they will be encoded in the file.
> 
> I wrote https://github.com/GMOD/bbi-js/ and https://github.com/GMOD/bed-js to try to solve these problems. Note that even these parsers do not go so far as to parse the features out into exons etc as that is sort of application dependent how you represent your gene features but it is just an idea

Yeah, the autoSQL is a good point! It would be nice to parse that and always be able to decode the data. I feel like that's a much larger change though. And, you also won't get any type information (which if you know the format ahead of time is useful). I could imagine that if/when we add support for parsing based on the autoSQL format, these parse functions can be augmented to *verify* that we are reading a file in the correct format first. In the mean time, I do also want to point out that this is compatible with no-op decoding the rest string and parsing separately.

Cool projects! I'll have to check them out.

**Comment by cmdcolin** [comment]
Cool, it is a good benefit. Typescript has been interesting to use in our codebases too...anyways if this module allows you to do-your-own-rest-parsing and can get the autosql that seems good too

**Comment by zoldello** [comment]
@jackh726  Thanks for the great feedback. I picked up some new things from you. You are awesome! Also, thanks for adding this contribution. It is very helpful and allows us in ENCODE-DCC to add new features.

**Pull request closed without merging**